File: zig/day06/guard_map.zig
const std = @import("std");

const Line = []u8;
const GuardMap = struct {
    map: []Line,
    height: usize,
    width: usize,
    pub fn init(map: []Line) GuardMap {
        return .{
            .map = map,
            .height = map.len,
            .width = map[0].len,
        };
    }
    pub fn deinit(self: *GuardMap, allocator: std.mem.Allocator) void {
        for (self.map) |line| {
            allocator.free(line);
        }
        allocator.free(self.map);
    }
    pub fn clone(self: *const GuardMap, allocator: std.mem.Allocator) !GuardMap {
        var clone_map = try allocator.alloc(Line, self.height);
        for (self.map, 0..) |line, row| {
            clone_map[row] = try allocator.alloc(u8, self.width);
            std.mem.copyForwards(u8, clone_map[row], line);
        }
        return GuardMap.init(clone_map);
    }
    pub fn copy(self: *const GuardMap, src: GuardMap) void {
        for (self.map, 0..) |line, row| {
            std.mem.copyForwards(u8, line, src.map[row]);
        }
    }
};

pub fn parseGuardMapFile(allocator: std.mem.Allocator, file_name: []const u8) !GuardMap {
    const file = try std.fs.cwd().openFile(file_name, .{});
    defer file.close();

    var buf_reader = std.io.bufferedReader(file.reader());
    const in_stream = buf_reader.reader();

    var lines_list = std.ArrayList(Line).init(allocator);
    defer lines_list.deinit();

    var lineBuf: [1024]u8 = undefined;
    while (try in_stream.readUntilDelimiterOrEof(&lineBuf, '\n')) |readLineBuf| {
        const line = try allocator.alloc(u8, readLineBuf.len);
        std.mem.copyForwards(u8, line, readLineBuf);
        try lines_list.append(line);
    }

    return GuardMap.init(try lines_list.toOwnedSlice());
}

test parseGuardMapFile {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    std.debug.print("day06/parseGuardMapFile\n", .{});
    std.debug.print("\tread test input file\n", .{});
    const guard_map = try parseGuardMapFile(allocator, "data/day06/test.txt");
    std.debug.print("\tcheck map\n", .{});
    for (&[_][]const u8{
        "....#.....",
        ".........#",
        "..........",
        "..#.......",
        ".......#..",
        "..........",
        ".#..^.....",
        "........#.",
        "#.........",
        "......#...",
    }, 0..) |expected, row| {
        std.debug.print("\t  {s}\n", .{expected});
        try std.testing.expectEqualStrings(expected, guard_map.map[row]);
    }
}

const GuardPosition = struct {
    direction: enum { UP, DOWN, LEFT, RIGHT },
    row: usize,
    col: usize,
};

fn extractInitialGuardPosition(guard_map: GuardMap) GuardPosition {
    for (guard_map.map, 0..) |line, row| {
        for (line, 0..) |char, col| {
            if (char == '^') {
                return .{
                    .direction = .UP,
                    .row = row,
                    .col = col,
                };
            }
        }
    } else unreachable;
}

const Obstruction = struct {
    row: usize,
    col: usize,
};

test extractInitialGuardPosition {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    std.debug.print("day06/extractInitialGuardPosition\n", .{});
    std.debug.print("\tread test input file\n", .{});
    const guard_map = try parseGuardMapFile(allocator, "data/day06/test.txt");
    std.debug.print("\tcheck initial position\n", .{});
    try std.testing.expectEqual(GuardPosition{
        .direction = .UP,
        .row = 6,
        .col = 4,
    }, extractInitialGuardPosition(guard_map));
}

const GuardPath = struct {
    positions: std.ArrayList(GuardPosition),
    pub fn init(allocator: std.mem.Allocator) GuardPath {
        return .{
            .positions = std.ArrayList(GuardPosition).init(allocator),
        };
    }
    pub fn deinit(self: *GuardPath) void {
        self.positions.deinit();
    }
    pub fn addPosition(self: *GuardPath, position: GuardPosition) !void {
        for (self.positions.items) |pos| {
            if (pos.row == position.row and pos.col == position.col and pos.direction == position.direction) {
                return error.LoopDetected;
            }
        }
        try self.positions.append(position);
    }
};

fn guardIsExiting(guard_map: GuardMap, guard_position: GuardPosition) bool {
    switch (guard_position.direction) {
        .UP => {
            if (guard_position.row == 0) return true;
        },
        .DOWN => {
            if (guard_position.row == guard_map.height - 1) return true;
        },
        .LEFT => {
            if (guard_position.col == 0) return true;
        },
        .RIGHT => {
            if (guard_position.col == guard_map.width - 1) return true;
        },
    }
    return false;
}

fn advanceGuard(guard_position: GuardPosition) GuardPosition {
    var next_position = guard_position;
    switch (guard_position.direction) {
        .UP => {
            next_position.row -= 1;
        },
        .DOWN => {
            next_position.row += 1;
        },
        .LEFT => {
            next_position.col -= 1;
        },
        .RIGHT => {
            next_position.col += 1;
        },
    }
    return next_position;
}

fn turnGuardRight(guard_position: GuardPosition) GuardPosition {
    var next_position = guard_position;
    next_position.direction = switch (guard_position.direction) {
        .UP => .RIGHT,
        .DOWN => .LEFT,
        .LEFT => .UP,
        .RIGHT => .DOWN,
    };
    return next_position;
}

fn moveGuardUntilExit(allocator: std.mem.Allocator, guard_map: GuardMap, start_position: GuardPosition) ![]GuardPosition {
    var guard_path = GuardPath.init(allocator);
    defer guard_path.deinit();

    var guard_position = start_position;
    try guard_path.addPosition(start_position);
    while (true) {
        if (guardIsExiting(guard_map, guard_position)) break;
        const next_position = advanceGuard(guard_position);
        if (guard_map.map[next_position.row][next_position.col] == '#') {
            guard_position = turnGuardRight(guard_position);
            continue;
        }
        guard_position = next_position;
        try guard_path.addPosition(guard_position);
    }
    return try guard_path.positions.toOwnedSlice();
}

fn checkLoop(allocator: std.mem.Allocator, guard_map: GuardMap, start_position: GuardPosition) !bool {
    var guard_path = GuardPath.init(allocator);
    defer guard_path.deinit();

    var guard_position = start_position;
    try guard_path.addPosition(start_position);
    while (true) {
        if (guardIsExiting(guard_map, guard_position)) return false;
        const next_position = advanceGuard(guard_position);
        if (guard_map.map[next_position.row][next_position.col] == '#') {
            guard_position = turnGuardRight(guard_position);
            continue;
        }
        guard_position = next_position;
        guard_path.addPosition(guard_position) catch |err| {
            if (err == error.LoopDetected) return true;
        };
    } else unreachable;
}

fn findAllPossibleLoopObstructions(allocator: std.mem.Allocator, guard_map: GuardMap, start_position: GuardPosition, visited_positions: []GuardPosition) ![]Obstruction {
    var possible_loop_obstructions = std.ArrayList(Obstruction).init(allocator);
    defer possible_loop_obstructions.deinit();
    const new_guard_map = try guard_map.clone(allocator);
    for (visited_positions[1..]) |position| {
        new_guard_map.copy(guard_map);
        new_guard_map.map[position.row][position.col] = '#';

        const has_loop = try checkLoop(allocator, new_guard_map, start_position);
        if (has_loop) {
            std.debug.print("possible loop position: row={d} col={d} direction={s}\n", .{ position.row, position.col, @tagName(position.direction) });
            const existing: ?Obstruction = for (possible_loop_obstructions.items) |obstruction| {
                if (obstruction.row == position.row and obstruction.col == position.col) {
                    break obstruction;
                }
            } else null;
            if (existing) |_| {
                std.debug.print("-- already exists\n", .{});
            } else {
                try possible_loop_obstructions.append(Obstruction{ .row = position.row, .col = position.col });
            }
        }
    }
    return try possible_loop_obstructions.toOwnedSlice();
}

fn markGuardPosition(visited_map: GuardMap, guard_position: GuardPosition) void {
    visited_map.map[guard_position.row][guard_position.col] = switch (guard_position.direction) {
        .UP => '^',
        .DOWN => 'v',
        .LEFT => '<',
        .RIGHT => '>',
    };
}

pub fn visitGuardMap(allocator: std.mem.Allocator, guard_map: GuardMap) !GuardMap {
    const start_position = extractInitialGuardPosition(guard_map);

    const visited_positions = try moveGuardUntilExit(allocator, guard_map, start_position);
    defer allocator.free(visited_positions);

    const possible_loop_obstructions = try findAllPossibleLoopObstructions(allocator, guard_map, start_position, visited_positions);
    defer allocator.free(possible_loop_obstructions);
    std.debug.print("possible loop obstructions: {}\n", .{possible_loop_obstructions.len});

    const visited_map = try guard_map.clone(allocator);
    for (visited_positions) |position| {
        markGuardPosition(visited_map, position);
    }
    return visited_map;
}

pub fn countVisitedPositions(visited_map: GuardMap) usize {
    var count: usize = 0;
    for (visited_map.map) |line| {
        for (line) |char| {
            if (char == '^' or char == 'v' or char == '<' or char == '>') {
                count += 1;
            }
        }
    }
    return count;
}

test moveGuardUntilExit {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    std.debug.print("day06/parseGuardMapFile\n", .{});
    std.debug.print("\tread test input file\n", .{});
    const guard_map = try parseGuardMapFile(allocator, "data/day06/test.txt");
    std.debug.print("\tmove guard until exit\n", .{});
    const visited_map = try visitGuardMap(allocator, guard_map);
    for (&[_][]const u8{
        "....#.....",
        "....^>>>>#",
        "....^...v.",
        "..#.^...v.",
        "..^>>>>#v.",
        "..^.^.v.v.",
        ".#<<<<v<v.",
        ".^>>>>>>#.",
        "#<<<<<vv..",
        "......#v..",
    }, 0..) |expected, row| {
        std.debug.print("\t  {s}\n", .{expected});
        try std.testing.expectEqualStrings(expected, visited_map.map[row]);
    }
    std.debug.print("\tcount visited positions\n", .{});
    try std.testing.expectEqual(41, countVisitedPositions(visited_map));
}

File: js/day06/guardMap.js
const fs = require('node:fs');
const { performance } = require('node:perf_hooks');

module.exports = {
  parseGuardMapFile,
  getInitialPosition,
  guardIsExiting,
  advanceGuard,
  turnGuardRight,
  traceGuardPath,
  findAllPossibleLoopObstructions,
  visitedMap,
};

function parseGuardMapFile({ fileName }) {
  const data = fs.readFileSync(fileName, 'utf8');
  const map = _.chain(data).split('\n').reject(_.isEmpty).value();
  return {
    guardMap: {
      map,
      height: _.size(map),
      width: _.chain(map).first().size().value(),
    },
  };
}

function getInitialPosition({ guardMap }) {
  for (const row in guardMap.map) {
    for (const col in guardMap.map[row]) {
      if (guardMap.map[row][col] === '^') {
        return {
          direction: 'up',
          row: Number(row),
          col: Number(col),
        };
      }
    }
  }
}

function guardIsExiting({ guardMap, guardPosition }) {
  return _.get(
    {
      up: guardPosition.row === 0,
      down: guardPosition.row === guardMap.height - 1,
      left: guardPosition.col === 0,
      right: guardPosition.col === guardMap.width - 1,
    },
    guardPosition.direction
  );
}

function advanceGuard({ guardPosition }) {
  return _.get(
    {
      up: { ...guardPosition, row: guardPosition.row - 1 },
      down: { ...guardPosition, row: guardPosition.row + 1 },
      left: { ...guardPosition, col: guardPosition.col - 1 },
      right: { ...guardPosition, col: guardPosition.col + 1 },
    },
    guardPosition.direction
  );
}

function turnGuardRight({ guardPosition }) {
  return _.get(
    {
      up: { ...guardPosition, direction: 'right' },
      down: { ...guardPosition, direction: 'left' },
      left: { ...guardPosition, direction: 'up' },
      right: { ...guardPosition, direction: 'down' },
    },
    guardPosition.direction
  );
}

function traceGuardPath({ guardMap }) {
  const initialPosition = getInitialPosition({ guardMap });
  let guardPosition = initialPosition;
  const path = [initialPosition];
  while (true) {
    if (guardIsExiting({ guardMap, guardPosition })) {
      break;
    }
    const nextPosition = advanceGuard({ guardPosition });
    if (guardMap.map[nextPosition.row][nextPosition.col] === '#') {
      guardPosition = turnGuardRight({ guardPosition });
      continue;
    }
    guardPosition = nextPosition;
    path.push(guardPosition);
  }
  return path;
}

function visitedMap({ guardMap, guardPath }) {
  const visitedMarkers = _.chain(guardPath)
    .reverse()
    .uniqBy(({ row, col }) => `${row},${col}`)
    .groupBy('row')
    .mapValues((v) =>
      _.chain(v)
        .keyBy('col')
        .mapValues('direction')
        .mapValues((direction) =>
          _.get({ up: '^', down: 'v', right: '>', left: '<' }, direction)
        )
        .value()
    )
    .value();
  return _.map(guardMap.map, (line, row) =>
    _.map(line, (char, col) => _.get(visitedMarkers, [row, col], char)).join('')
  );
}

function setMapMarker({ guardMap, row, col, marker }) {
  const newGuardMap = _.cloneDeep(guardMap);
  newGuardMap.map[row] =
    `${newGuardMap.map[row].substring(0, col)}${marker}${newGuardMap.map[row].substring(col + 1)}`;
  return newGuardMap;
}

function measure() {
  performance.measure('checkExitDuration', 'start', 'checkExit');
  performance.measure('checkTurnDuration', 'checkExit', 'checkTurn');
  performance.measure('checkLoopDuration', 'checkTurn', 'checkLoop');
  performance.measure('checkEndDuration', 'checkLoop', 'end');
}

function perf() {
  inspect({
    checkExitDuration: _.sumBy(
      performance.getEntriesByName('checkExitDuration'),
      'duration'
    ),
    checkTurnDuration: _.sumBy(
      performance.getEntriesByName('checkTurnDuration'),
      'duration'
    ),
    checkLoopDuration: _.sumBy(
      performance.getEntriesByName('checkLoopDuration'),
      'duration'
    ),
    checkEndDuration: _.sumBy(
      performance.getEntriesByName('checkEndDuration'),
      'duration'
    ),
  });
}

function checkLoop({ guardMap }) {
  const initialPosition = getInitialPosition({ guardMap });
  let guardPosition = initialPosition;
  const path = [initialPosition];
  const start = new Date();
  let loop = false;
  while (true) {
    //performance.mark('start');
    if (guardIsExiting({ guardMap, guardPosition })) {
      inspect({ EXIT: guardPosition, duration: new Date() - start });
      //performance.mark('checkExit');
      break;
    }
    //performance.mark('checkExit');
    const nextPosition = advanceGuard({ guardPosition });
    if (guardMap.map[nextPosition.row][nextPosition.col] === '#') {
      guardPosition = turnGuardRight({ guardPosition });
      //performance.mark('checkTurn');
      continue;
    }
    //performance.mark('checkTurn');
    guardPosition = nextPosition;
    if (
      _.some(
        path,
        (p) =>
          p.row === guardPosition.row &&
          p.col === guardPosition.col &&
          p.direction === guardPosition.direction
      )
    ) {
      inspect({ LOOP: guardPosition, duration: new Date() - start });
      loop = true;
      //performance.mark('checkLoop');
      break;
    }
    //performance.mark('checkLoop');
    path.push(guardPosition);
    //performance.mark('end');
    //measure();
  }
  //measure();
  //perf();
  return loop;
}

function findAllPossibleLoopObstructions({ guardMap, guardPath }) {
  const initialPosition = getInitialPosition({ guardMap });
  return _.chain(guardPath)
    .reject(
      ({ row, col }) =>
        row === initialPosition.row && col === initialPosition.col
    )
    .map(({ row, col }) => ({ row, col }))
    .uniqBy(({ row, col }) => `${row},${col}`)
    .filter(({ row, col }, index) => {
      inspect({ index });
      const newGuardMap = setMapMarker({ guardMap, row, col, marker: '#' });
      return checkLoop({ guardMap: newGuardMap });
    })
    .orderBy(['row', 'col'])
    .value();
}
